packages/sensors_plus/tizen/src/sensors_plus_plugin.cc

```diff
@@ -3,221 +3,89 @@
 
-#include <functional>
-#include <memory>
-#include <string>
+#include "device_sensor.h"
 
-#include "log.h"
+typedef flutter::EventChannel<flutter::EncodableValue> FlEventChannel;
+typedef flutter::EventSink<flutter::EncodableValue> FlEventSink;
+typedef flutter::StreamHandler<flutter::EncodableValue> FlStreamHandler;
+typedef flutter::StreamHandlerError<flutter::EncodableValue>
+    FlStreamHandlerError;
 
-class Listener {
+class DeviceSensorStreamHandler : public FlStreamHandler {
  public:
-  Listener(
-      std::unique_ptr<flutter::EventSink<flutter::EncodableValue>> &&event_sink)
-      : event_sink_(std::move(event_sink)) {}
-
-  bool Init(sensor_type_e type) {
-    sensor_h sensor;
-    int ret = sensor_get_default_sensor(type, &sensor);
-    if (ret != SENSOR_ERROR_NONE) {
-      LOG_ERROR("%s", get_error_message(ret));
-      return false;
-    }
-    ret = sensor_create_listener(sensor, &listener_);
-    if (ret != SENSOR_ERROR_NONE) {
-      LOG_ERROR("%s", get_error_message(ret));
-      return false;
-    }
-    return true;
-  }
-
-  void Listen() {
-    if (is_listening_) {
-      LOG_ERROR("Already listening!");
-      return;
-    }
-
-    int ret = sensor_listener_set_event_cb(
-        listener_, 60,
-        [](sensor_h sensor, sensor_event_s *event, void *user_data) {
-          Listener *s = (Listener *)user_data;
-          std::vector<double> list;
-          for (int i = 0; i < event->value_count; i++) {
-            list.push_back(event->values[i]);
-          }
-          flutter::EncodableValue value(list);
-          s->event_sink_->Success(value);
-        },
-        this);
-
-    if (ret != SENSOR_ERROR_NONE) {
-      LOG_ERROR("%s", get_error_message(ret));
-      return;
-    }
-    ret = sensor_listener_start(listener_);
-    if (ret != SENSOR_ERROR_NONE) {
-      LOG_ERROR("%s", get_error_message(ret));
-      return;
-    }
-
-    is_listening_ = true;
-  }
+  DeviceSensorStreamHandler(SensorType type) : sensor_(type) {}
 
-  void Cancel() {
-    if (!is_listening_) {
-      LOG_ERROR("Already canceled!");
-      return;
-    }
-    int ret = sensor_listener_stop(listener_);
-    if (ret != SENSOR_ERROR_NONE) {
-      LOG_ERROR("%s", get_error_message(ret));
-      return;
-    }
-
-    ret = sensor_listener_unset_event_cb(listener_);
-    if (ret != SENSOR_ERROR_NONE) {
-      LOG_ERROR("%s", get_error_message(ret));
-      return;
+ protected:
+  std::unique_ptr<FlStreamHandlerError> OnListenInternal(
+      const flutter::EncodableValue *arguments,
+      std::unique_ptr<FlEventSink> &&events) override {
+    events_ = std::move(events);
+
+    SensorEventReceivedCallback callback =
+        [this](SensorEvent sensor_event) -> void {
+      events_->Success(flutter::EncodableValue(sensor_event));
+    };
+    if (!sensor_.StartListen(callback)) {
+      return std::make_unique<FlStreamHandlerError>(
+          std::to_string(sensor_.GetLastError()), sensor_.GetLastErrorString(),
+          nullptr);
```

Comment:
This currently results in an app crash. see: https://github.com/flutter-tizen/plugins/pull/351#discussion_r847005275